File: contest/2022-10-12 (2)/资料/10-02/Src/hzg/DerProze.pas
program DerProze;
var i,year,month,day:longint;

function nian(x:longint):longint;
begin
if x mod 100<>0 then
        begin
        if x mod 4<>0 then exit(365) else exit(366);
        end
else
        begin
        if x mod 400<>0 then exit(365) else exit(366);
        end;
end;

function yue(x,y:longint):longint;
var r:longint;
begin
if(x=1)or(x=3)or(x=5)or(x=7)or(x=8)or(x=10)or(x=12)then exit(31)
else
if (x=4)or(x=6)or(x=9)or(x=11) then exit(30)
else
if nian(y)=365 then exit(28)
else exit(29);
end;

function find(x,y:longint):longint;
var r,k,l:longint;
begin
if y=2 then
        begin
        k:=x mod 7;
        l:=5;
        for r:=1 to k do l:=(l+1) mod 7;
        exit(l);
        end
else
        begin
        k:=x mod 7;
        l:=5;
        for r:=1 to k do
                begin
                if l=0 then l:=6 else dec(l);
                end;
        exit(l);
        end;
end;

function cal(x,y,z:longint):longint;
var r,k,boo,ans,std:longint;
begin
if x<2011 then
        begin
        ans:=0;
        for r:=y+1 to 12 do inc(ans,yue(r,x));
        inc(ans,yue(y,x)-z);
        for r:=x+1 to 2010 do inc(ans,nian(r));
        for r:=1 to 9 do inc(ans,yue(r,2011));
        k:=find(ans,1);
        exit(k);
        end
else
if x>2011 then
        begin
        ans:=0;
        for r:=10 to 12 do inc(ans,yue(r,2011));
        for r:=2012 to x-1 do inc(ans,nian(r));
        for r:=1 to y-1 do inc(ans,yue(r,x));
        inc(ans,z);
        k:=find(ans,2);
        exit(k);
        end
else
        begin
        std:=0;ans:=0;
        for r:=1 to 9 do inc(std,yue(r,x));
        for r:=1 to y-1 do inc(ans,yue(r,x));
        inc(ans,z);
        boo:=1;
        if ans>std then boo:=2;
        ans:=abs(ans-std);
        k:=find(ans,boo);
        exit(k);
        end;
end;

procedure shuchu(x:longint);
begin
if x=1 then writeln('monday');
if x=2 then writeln('tuesday');
if x=3 then writeln('wednesday');
if x=4 then writeln('thursday');
if x=5 then writeln('friday');
if x=6 then writeln('saturday');
if x=0 then writeln('sunday');
end;

begin
assign(input,'DerProze.in');reset(input);
assign(output,'DerProze.out');rewrite(output);
readln(year,month,day);
i:=cal(year,month,day);
shuchu(i);
close(input);close(Output);
end.

File: contest/2022-10-12 (2)/资料/10-02/Src/zjc/DerProze.cpp
#include<stdio.h>
int n,m,t,ans;
inline int year(int x)
{
    if((x%4==0)&&((x%100!=0)||(x%400==0))) return 366;
    return 365;
}
inline int month(int x)
{
    if((x==1)||(x==3)||(x==5)||(x==7)||(x==8)||(x==10)||(x==12)) return 31;
    if((x==4)||(x==6)||(x==9)||(x==11)) return 30;
    if((n%4==0)&&((n%100!=0)||(n%400==0))) return 29;
    return 28;
}
int main()
{
    freopen("DerProze.in","r",stdin);
    freopen("DerProze.out","w",stdout);
    int i,j,k;
    scanf("%d%d%d",&n,&m,&t);
    if(n>=2007) for(i=2007;i<n;i++) ans+=year(i);
    else for(i=2006;i>=n;i--) ans-=year(i);
    for(i=1;i<m;i++) ans+=month(i);
    ans+=t;
    ans=(ans%7+7)%7;
    if(ans==1) printf("monday\n");
    if(ans==2) printf("tuesday\n");
    if(ans==3) printf("wednesday\n");
    if(ans==4) printf("thursday\n");
    if(ans==5) printf("friday\n");
    if(ans==6) printf("saturday\n");
    if(ans==0) printf("sunday\n");
    return 0;
}
    
